python/tvm/relay/backend/contrib/ethosu/codegen.py

```diff
@@ -52,29 +60,147 @@ def constant_updater(expr, symbol):  # pylint: disable=unused-argument
     return dict()
 
 
+class LayoutOptimization(ExprMutator):
+    """A pass to optimize the layout of NPU operations. If both the
+    producer and consumer of a tensor are NPU operators, then the
+    layout is converted from NHWC to NHCWB16.
+    """
+
+    def __init__(self):
+        self.children = {}
+        self.optimize_op = {
+            "contrib.ethosu.conv2d": ethosu_op.ethosu_conv2d,
+            "contrib.ethosu.depthwise_conv2d": ethosu_op.ethosu_depthwise_conv2d,
+            "contrib.ethosu.pooling": ethosu_op.ethosu_pooling,
+            "contrib.ethosu.binary_elementwise": ethosu_op.ethosu_binary_elementwise,
+        }
+
+        super().__init__()
+
+    def alter_ethosu_op_layout(self, call: tvm.relay.expr.Call) -> tvm.relay.expr.Call:
+        """Alter the input and output layouts of an NPU operation if needed.
+        Input layout is only altered if the producing operation is an NPU
+        operation. Likewise, the output layout is only altered if the consuming
+        operation is an NPU operation.
+
+        Parameters
+        ----------
+        call : tvm.relay.expr.Call
+            The call pointing to an NPU operation that will be checked if
+            the layout needs altering.
+
+        Returns
+        -------
+        new_call : tvm.relay.expr.Call
+            New call with altered layouts.
+        """
+        assert isinstance(call.attrs, tvm.ir.Attrs), (
+            f"The attributes for operator '{call.op.name}' could not be "
+            "found. Did you register the relay.attrs.Ethosu<opname>Attrs "
+            "object in python api?"
+        )
+
+        new_attrs = dict(call.attrs)
+        parents = []
+
+        # Check if we can rewrite the input layouts
+        layout_count = 0
```

Comment:
Nit: I think name `layout_count` is slightly misleading since we are essentially counting inputs, not layouts...